pkg/bucketaccess/bucketaccess_controller.go

```diff
@@ -350,10 +352,18 @@ func (bal *BucketAccessListener) deleteBucketAccessOp(ctx context.Context, bucke
 		klog.V(3).ErrorS(err, "Failed to fetch bucket", "bucket", bucketClaim.Status.BucketName)
 		return fmt.Errorf("failed to fetch bucket: %w", err)
 	}
-
+	bucketAccessClassName := bucketAccess.Spec.BucketAccessClassName
+	bucketAccessClass, err := bal.bucketAccessClasses().Get(ctx, bucketAccessClassName, metav1.GetOptions{})
+	if kubeerrors.IsNotFound(err) {
+		return bal.recordError(bucketAccess, v1.EventTypeWarning, events.FailedGrantAccess, err)
+	} else if err != nil {
+		klog.V(3).ErrorS(err, "Failed to fetch bucketAccessClass", "bucketAccessClass", bucketAccessClassName)
+		return fmt.Errorf("failed to fetch BucketAccessClass: %w", err)
+	}
```

Comment:
I have seen issues with this approach in CSI drivers. This leads to drivers implementing code that looks to the StorageClass (approx equivalent to BucketAccessClass), which is mutable and can change over time (or be deleted). This then results in a situation where the driver is unable to proceed with the latest operation.

Instead, the best pattern I have seen is to have the StorageClass data (including opaque params) copied to the PV so that the PV can be managed regardless of whether the StorageClass has been mutated.

I would suggest that the resolution follow the suggested pattern rather than adding the class name, which in my experience encourages inconsistent expectations.

@xing-yang or @shanduur do you have other thoughts?

Reply:
Hi @BlaineEXE ,

Thanks for your suggestion.

Certainly agree with your point here. 
As per your suggestion, there's 2 ways (which I can think of), through which Parameters can be made available to the 'DriverRevokeBucketAccess' API.

- **Approach-1**: Take Parameter from the Bucket CR spec

   Pros:
   1) Given that the Bucket is already retrieved within the method, makes it easy to fetch parameters field from its spec.
     `req := &cosi.DriverRevokeBucketAccessRequest{
		BucketId:            bucket.Status.BucketID,
		AccountId:           bucketAccess.Status.AccountID,
		RevokeAccessContext: bucket.Spec.Parameters,
	}`
   https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar/blob/80979e8992a6a2b2166f3ff1e7d39b4ab03f045c/pkg/bucketaccess/bucketaccess_controller.go#L348
   
   Cons:
   1) This would imply that the Parameters passed in BucketClass & BucketAccessClass should always be same (which it should not be).
   2) Given, at the time of BucketAccess Grant, Parameter is retrieved & passed from the BucketAccessClass. The 
      same should be followed for Revoking BucketAccess as well.
      https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar/blob/80979e8992a6a2b2166f3ff1e7d39b4ab03f045c/pkg/bucketaccess/bucketaccess_controller.go#L170

-  **Approach-2**: Save Parameter in BucketAccess CR spec at the time of creation
   
   Pros:
   1) Same approach seems to be followed for Bucket CR
      https://github.com/kubernetes-sigs/container-object-storage-interface-api/blob/8f3e219335924b998c49c03056d61674361bd132/crds/objectstorage.k8s.io_buckets.yaml#L97
    2) At the time of BucketAccess CR creation, parameter would be retrieved from the BucketAccessClass CR & stored in its spec.
   3) Parameter can be directly retrieved from the BucketAccess CR spec 
     `req := &cosi.DriverRevokeBucketAccessRequest{
		BucketId:            bucket.Status.BucketID,
		AccountId:           bucketAccess.Status.AccountID,
		RevokeAccessContext: bucketAccess.Spec.Parameters,
	}`
    4) Parameters in BucketAccessClass can be different & independent of BucketClass CR ( which seems to be the right way).
        
    Cons:
    1) Changes has to be made in BucketAccess CRD, to support parameter field in its spec.
        https://github.com/kubernetes-sigs/container-object-storage-interface-api/blob/8f3e219335924b998c49c03056d61674361bd132/crds/objectstorage.k8s.io_bucketaccesses.yaml#L34
         
I am more in-favor of 'Approach-2', but kindly do guide me if there's another way to retrieve 'Parameters' before 'DriverRevokeBucketAccess' API is invoked.

Will wait for your, suggestion & incorporate changes accordingly.

Reply:
I think approach 2 is also the more correct one. That means a change to the API in addition to the sidecar.

The next question I have is a matter of SIG/KEP process. With a CRD change, we may need to change the KEP. @xing-yang is it okay to make this change (what I consider a minor change) to the repos, or should we update the KEP first?

Reply:
Hi @BlaineEXE , @xing-yang,
 Any update on this..?

Reply:
If I can add my 5-cents, I am also in favor of the approach 2. Unfortunately, I think those changes will be required to be added to the KEP for v1alpha2, as this is pretty major breaking change in the controller behavior.